test: test with v3.0 (preparation for v3.1)

gyuho opened the pull request:
For https://github.com/coreos/etcd/pull/6309#issuecomment-243523255.

Confirmed that this fails without Xiang's patch

```
2016-08-30 16:47:58.995114 C | etcdmain: database file (/tmp/testname1.etcd397787835/member/snap/db index 0) does not match with snapshot (index 16).
2016-08-30 16:47:58.998345 C | etcdmain: database file (/tmp/testname2.etcd766704350/member/snap/db index 0) does not match with snapshot (index 16).
2016-08-30 16:47:59.000010 C | etcdmain: database file (/tmp/testname0.etcd443864012/member/snap/db index 0) does not match with snapshot (index 16).
```

/cc @xiang90 @heyitsanthony

xiang90 commented on e2e/etcd_release_upgrade_test.go:
```diff
@@ -86,3 +87,58 @@ func TestReleaseUpgrade(t *testing.T) {
 	}
 }
+
+func TestReleaseUpgradeV2ToV3(t *testing.T) {
```
i do not think we need the actual binary. this test is good in general for ensuring upgrade working.

gyuho commented:
Currently we only test with latest `v3.0.x` releases. Will change it to use only one binary as Xiang suggested.

gyuho commented:
@xiang90 PTAL. Changed it to use `v2.3.x`

xiang90 commented on e2e/etcd_release_upgrade_test.go:
```diff
@@ -37,7 +35,11 @@ func TestReleaseUpgrade(t *testing.T) {
 	copiedCfg.execPath = lastReleaseBinary
 	copiedCfg.snapCount = 3
-	copiedCfg.baseScheme = "unix" // to avoid port conflict
+
```
we should not change this test. we should add a new test called TestReleaseUpgradeWithSnapshot.

gyuho replied:
@xiang90 Sure, then we will need to download separate `v3.0.x` binary until we can drop `v2.7.x`?

xiang90 replied:
both tests should test 2.3.x -> 3.0.x.

xiang90 replied:
after we release 3.1, the test should test 3.0.x -> 3.1.x.

xiang90 commented on e2e/etcd_test.go:
```diff
@@ -141,6 +141,9 @@ type etcdProcessConfig struct {
 	isProxy bool
+
+	// TODO: deprecate this and related v2-handling after v3.1
+	isV2Migrate bool
```
isV2Cluster

xiang90 commented on e2e/etcd_test.go:
```diff
@@ -451,14 +472,17 @@ func (ep *etcdProcess) Stop() error {
 }
 
-func waitReadyExpectProc(exproc *expect.ExpectProcess, isProxy bool) error {
+func waitReadyExpectProc(exproc *expect.ExpectProcess, isProxy, isV2 bool) error {
```
waitReadExpecProc(exproc ..., readString string)

we should avoid bool arg when possible. we should try VERY HARD to not use two bools.

gyuho commented:
@xiang90 Now `TestReleaseUpgrade` fails, where we only write v3 keys. So `v2.3.7` with experimental feature on and with only v3 keys fails to migrate to `v3.x`. Can you take a look?

```
etcdmain: database file (/tmp/testname1.etcd433904263/member/snap/db index 13) does not match with snapshot (index 16)
```

gyuho commented:
Current `TestReleaseUpgrade` and the new `TestReleaseUpgradeWithSnapshot` are almost the same.

I think we should combine those two after `v3.1`.

xiang90 commented on e2e/etcd_release_upgrade_test.go:
```diff
@@ -86,3 +91,55 @@ func TestReleaseUpgrade(t *testing.T) {
 	}
 }
+
+func TestReleaseUpgradeWithSnapshot(t *testing.T) {
```
This test case is not exactly what i described here: https://github.com/coreos/etcd/pull/6309#issuecomment-243523255

we need to restart the new cluster at the end of test.

gyuho replied:
How about 

```go
	for i := range epc.procs {
		if err := epc.procs[i].Stop(); err != nil {
			t.Fatalf("#%d: error closing etcd process (%v)", i, err)
		}
	}

	var wg sync.WaitGroup
	wg.Add(len(epc.procs))
	for i := range epc.procs {
		go func(i int) {
			epc.procs[i].cfg.execPath = binDir + "/etcd"
			epc.procs[i].cfg.keepDataDir = true
			epc.procs[i].cfg.isV2Cluster = false
			if err := epc.procs[i].Restart(); err != nil {
				t.Fatalf("error restarting etcd process (%v)", err)
			}
			wg.Done()
		}(i)
	}
	wg.Wait()
}
```

This restarts after killing the whole cluster.

xiang90 replied:
```
Set snap count to 10.
Write 50 keys.
upgrade cluster.
kill the upgraded cluster without doing any more v3 writes
restart the upgraded cluster and verify everything is OK
```

1. we should write 50 keys instead of 10 to ensure a snapshot is triggered
2. after the final restart, we need to check the cluster works

gyuho replied:
Yeah just fixed as it is.

`TestReleaseUpgrade` still fails, though.

xiang90 commented:
we should not enable experiment flag for v2 cluster. the v2 experimental cluster cannot upgrade to v3. we need to write v2 keys to trigger snapshot for v2 cluster, not v3 keys.

gyuho commented:
@xiang90 Right, I will switch to v2 keys then.

xiang90 commented:
@gyuho I did not know that we were testing 3.0.x to 3.0.x upgrading. If it requires too much effort, we can probably drop this pr. we are going to release 3.1 rc soon anyway...

gyuho commented:
@xiang90 Ok, just dropped v2 tests part. But for v3.1 releases, I change `test` script to download `v3.0.x`. 

Or should we just keep getting the latest releases, like `3.1`, `3.1.1`, ...?

Also dropped unnecessary function `stripSchema`.

xiang90 commented:
right before we branch out 3.1 we should test 3.0 -> 3.1. right before we branch out 3.2, we should test 3.1 -> 3.2.

gyuho commented:
Ok will just keep test script change.

xiang90 commented:
lgtm
